services/users.ts

```diff
@@ -83,36 +85,29 @@ export const queryUsers = async (
   }
 };
 
-export const updateUser = async (
-  req: NextApiRequest,
-  res: NextApiResponse<Omit<User, 'password'>>,
-  session: Session | null
-) => {
-  try {
-    if (!session) {
-      return NextResponse.json({ status: 403 });
-    }
-
-    const role = session.user.role;
-    let { id } = req.query;
-
-    // Normal user can only update their own account
-    if (role === UserRole.USER || role === UserRole.AUTHOR) {
-      id = session.user.id;
-    }
+export const updateUser = async (id: string, data: Partial<User>) => {
+  const session = await getSession();
 
-    const result = await prisma.user.update({
-      where: { id: id as string },
-      data: JSON.parse(req.body),
-    });
+  if (!session) {
+    throw new Error('Unauthorized request');
+  }
 
-    return NextResponse.json({
-      status: 200,
-      result: exclude(result, 'password'),
-    });
-  } catch (error) {
-    return NextResponse.json({ status: 500 });
+  if (!id || !data) {
+    throw new Error('id and data update must be request');
+  }
+  const role = session.user.role;
+  // Normal user can only update their own account
+  if (role === UserRole.USER || role === UserRole.AUTHOR) {
+    id = session.user.id;
   }
+  const result = await prisma.user.update({
+    where: { id: id as string },
+    data: {
+      ...data,
+      preferences: data.preferences as Prisma.JsonObject,
```

Comment:
double check to make sure it will be deep merged

Reply:
Với nhiều lần research thì em chọn cách cập nhật theo cách lấy giá trị preferences hiện tại mới nhất từ database rồi cập nhật vô. (Chat GPT chỉ vậy 😅)